tools/build_langserver/langserver/request_store.go

```diff
@@ -0,0 +1,57 @@
+package langserver
+
+import (
+	"context"
+	"sync"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+type requestStore struct {
+	mu       sync.Mutex
+	requests map[jsonrpc2.ID]request
+}
+
+// Perhaps later we can store more things in the request we might want to use
+type request struct {
+	id     string
+	cancel func()
+}
+
+func (rs *requestStore) Store(ctx context.Context, req *jsonrpc2.Request) context.Context {
+	ctx, cancel := context.WithCancel(ctx)
+
+	rs.mu.Lock()
+	if rs.requests == nil {
+		rs.requests = make(map[jsonrpc2.ID]request)
+	}
+	rs.mu.Unlock()
+
+	rs.mu.Lock()
```

Comment:
don't unlock the mutex then lock it again...

Reply:
o.o, I changed my code, and your comment is still here o.o